config/initializers/secure_headers.rb

```diff
@@ -0,0 +1,42 @@
+SecureHeaders::Configuration.default do |config|
+  config.cookies = {
+    secure: true,
+    httponly: true
+  }
+  # Add "; preload" and submit the site to hstspreload.org for best protection.
+  config.hsts = "max-age=#{20.years.to_i}; includeSubdomains"
+  config.x_frame_options = "SAMEORIGIN"
+  config.x_content_type_options = "nosniff"
+  config.x_xss_protection = "1; mode=block"
+  config.x_permitted_cross_domain_policies = "none"
+  config.referrer_policy = "origin-when-cross-origin"
+
+  # Valid for assets
+  assets_src = ["'self'"]
+  asset_host = OpenProject::Configuration.rails_asset_host
+  assets_src << asset_host if asset_host.present?
+
+  config.csp = {
+    preserve_schemes: true,
+
+    # Fallback when no value is defined
+    default_src: %w(https: 'self'),
+    # Allowed uri in <base> tag
+    base_uri: %w('self'),
+
+    # Allow fonts from self, asset host, or DATA uri
+    font_src: assets_src + %w(data:),
+    # Form targets can only be self
+    form_action: %w('self'),
+    # Allow iframe from vimeo (welcome video)
+    frame_src: %w(https://*.vimeo.com 'self'),
+    frame_ancestors: %w('self'),
+    # Allow images from anywhere
+    img_src: %w(* data:),
+    # Allow scripts from self (not inline, but)
+    # for now require unsafe-eval for Angular JIT
+    script_src: %w('self' 'unsafe-eval'),
+    # Allow unsafe-inline styles
+    style_src: assets_src + %w('unsafe-inline'),
```

Comment:
@oliverguenther Request for clarification: Isn't the purpose of all this SCP effort to disallow inline styles and JS? Why is it that we trust inline-styles?

Reply:
Yeah,  ideally, we can disable inline css as well. But that will kill all `<div style="display: none">..` and other things we have laying around. This is just not feasible at the moment and will blow up the PR by several order of magnitude.

That means that in theory, the cross-site styling attack vector is still open (e.g., a [CSS keylogger](https://css-tricks.com/css-keylogger/) would be possible.

We should try and use `hidden` or semantic classes to set initial styles when rendering from the backend. It's definitely possible, but hard.

The same goes for the `unsafe-eval` used for ng2 JIT. We can disable this as soon as we use the AOT build to compile all templates beforehand.

Reference:
https://stackoverflow.com/a/31759553
https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/Content-Security-Policy/style-src

Reply:
@wielinde Sadly, using Angular JIT needs unsave-eval, unlike AoT. https://github.com/angular/angular/issues/12958

Reply:
@oliverguenther are there plans to switch to AoT?

Reply:
@braiam Yes, we are going to switch before the angular upgrade goes into production (OpenProject 8.0.). There's been difficulties in bringing AOT and ng1 together so we'll need to see what and when that can happen.